### ISSUE TITLE
OpenSim.Data.MySQL.dll - Cannot Find

### ISSUE THREAD

**Issue by DJShocker** [HalcyonGrid/halcyon#44]
So I've pulled the repo using Visual Studio, built with no errors, but crashes:
 LOGS

> 
> > 2018-12-20 00:32:04,778 [1] INFO  OpenSim.Application - [HALCYON MAIN]: configured log4net using default Halcyon.exe.config
> 2018-12-20 00:32:04,795 [1] INFO  OpenSim.Application - Performing compatibility checks... 
> 2018-12-20 00:32:04,798 [1] INFO  OpenSim.Application - Environment is compatible.
> 
> 2018-12-20 00:32:04,816 [ConsoleThread] INFO  OpenSim.ConfigurationLoader - [CONFIG] Reading configuration settings
> 2018-12-20 00:32:04,817 [ConsoleThread] INFO  OpenSim.ConfigurationLoader - [CONFIG] Reading configuration file C:\Users\Administrator\Desktop\Halcyon\bin\Halcyon.ini
> 2018-12-20 00:32:04,856 [ConsoleThread] INFO  OpenSim.Framework.Servers.BaseOpenSimServer - [STARTUP]: Beginning startup processing
> 2018-12-20 00:32:04,859 [ConsoleThread] INFO  OpenSim.OpenSim - ====================================================================
> 2018-12-20 00:32:04,859 [ConsoleThread] INFO  OpenSim.OpenSim - ========================= STARTING HALCYON =========================
> 2018-12-20 00:32:04,860 [ConsoleThread] INFO  OpenSim.OpenSim - ====================================================================
> 2018-12-20 00:32:04,860 [ConsoleThread] INFO  OpenSim.OpenSim - [HALCYON MAIN]: Running in sandbox mode
> 2018-12-20 00:32:04,861 [ConsoleThread] INFO  OpenSim.OpenSim - GC: Server mode: False, Interactive
> 2018-12-20 00:32:04,873 [ConsoleThread] INFO  OpenSim.Region.Framework.StorageManager - [DATASTORE]: Attempting to load OpenSim.Data.MySQL.dll
> 2018-12-20 00:32:04,881 [ConsoleThread] ERROR OpenSim.Application - [APPLICATION]: 
> APPLICATION EXCEPTION DETECTED: System.UnhandledExceptionEventArgs
> 
> Exception: System.IO.FileNotFoundException: Could not load file or assembly 'file:///C:\Users\Administrator\Desktop\Halcyon\bin\OpenSim.Data.MySQL.dll' or one of its dependencies. The system cannot find the file specified.
> File name: 'file:///C:\Users\Administrator\Desktop\Halcyon\bin\OpenSim.Data.MySQL.dll'
>    at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
>    at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
>    at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
>    at System.Reflection.Assembly.LoadFrom(String assemblyFile)
>    at OpenSim.Region.Framework.StorageManager..ctor(String dllName, String connectionstring, String estateconnectionstring) in C:\Users\Administrator\source\repos\halcyon\OpenSim\Region\Framework\StorageManager.cs:line 61
>    at OpenSim.OpenSimBase.CreateStorageManager(String connectionstring, String estateconnectionstring) in C:\Users\Administrator\source\repos\halcyon\OpenSim\Base\OpenSimBase.cs:line 658
>    at OpenSim.OpenSimBase.CreateStorageManager() in C:\Users\Administrator\source\repos\halcyon\OpenSim\Base\OpenSimBase.cs:line 652
>    at OpenSim.Region.ClientStack.RegionApplicationBase.StartupSpecific() in C:\Users\Administrator\source\repos\halcyon\OpenSim\Region\ClientStack\RegionApplicationBase.cs:line 103
>    at OpenSim.OpenSimBase.StartupSpecific() in C:\Users\Administrator\source\repos\halcyon\OpenSim\Base\OpenSimBase.cs:line 201
>    at OpenSim.OpenSim.StartupSpecific() in C:\Users\Administrator\source\repos\halcyon\OpenSim\Base\OpenSim.cs:line 163
>    at OpenSim.Framework.Servers.BaseOpenSimServer.Startup() in C:\Users\Administrator\source\repos\halcyon\OpenSim\Framework\Servers\BaseOpenSimServer.cs:line 300
>    at OpenSim.Application.Main(String[] args) in C:\Users\Administrator\source\repos\halcyon\InWorldz\Halcyon\Application.cs:line 153
> 
> WRN: Assembly binding logging is turned OFF.
> To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
> Note: There is some performance penalty associated with assembly bind failure logging.
> To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
> 
> 
> Application is terminating: True

> *** Assembly Binder Log Entry  (12/20/2018 @ 12:36:31 AM) ***
> 
> The operation failed.
> Bind result: hr = 0x80070002. The system cannot find the file specified.
> 
> Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
> Running under executable  C:\Users\Administrator\Desktop\Halcyon\bin\Halcyon.exe
> --- A detailed error log follows. 
> 
> === Pre-bind state information ===
> LOG: Where-ref bind. Location = C:\Users\Administrator\Desktop\Halcyon\bin\OpenSim.Data.MySQL.dll
> LOG: Appbase = file:///C:/Users/Administrator/Desktop/Halcyon/bin/
> LOG: Initial PrivatePath = NULL
> LOG: Dynamic Base = NULL
> LOG: Cache Base = NULL
> LOG: AppName = Halcyon.exe
> Calling assembly : (Unknown).
> ===
> LOG: This bind starts in LoadFrom load context.
> WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
> LOG: Using application configuration file: C:\Users\Administrator\Desktop\Halcyon\bin\Halcyon.exe.Config
> LOG: Using host configuration file: 
> LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
> LOG: Attempting download of new URL file:///C:/Users/Administrator/Desktop/Halcyon/bin/OpenSim.Data.MySQL.dll.
> LOG: All probing URLs attempted and failed.

**Comment by kf6kjg** [comment]
That usually means you've got leftover files in your working copy from a previous edition.  

Firstly backup your config files:
* `UserServer_Config.xml`
* `server.p12`
* `server.crt`
* `MessagingServer_Config.xml`
* `Halcyon.ini`
* `GridServer_Config.xml`
* And the `Regions` folder

Please note that you might not have all of these files in your copy, and you might have others you'll want to keep.

Two ways to go at this point:
1. Delete the working copy and re-clone from server, or
2. Execute `git clean -xdf`

Either way you'll have a very sparkling clean working copy.

Then copy back in the files you backed up, execute `runprebuild.bat`, then open the solution in Visual Studio 2017, and you might have to build twice: sometimes NuGet doesn't get everything on the first go and you get a failed compile, not sure why.

**Comment by DJShocker** [comment]
Really confused as this was a fresh clean pull from GitHub. This is my first time attempting this

**Comment by kf6kjg** [comment]
Interesting.

I'm assuming that you ran `runprebuild.bat` otherwise nothing would have compiled.  When you compiled did you get any compilation errors in Visual Studio?

Looking closer at your errors I see you are running in standalone mode.  I've not tested that mode in quite a while: I run in grid mode every time just like I would when deploying to a real grid. Try editing your `Halcyon.ini` and set `gridmode = true`, then launch each service in order finishing with the region.  You can read some level of detail on that on the wiki: https://github.com/HalcyonGrid/halcyon/wiki/Configuration

Thus far I've not been able to reproduce the problem on my machine.  If the above doesn't fix it, could you give me a detailed blow-by-blow ELI5-style on how you compile, configure, and run it?

**Comment by DJShocker** [comment]
Well I got it out of that, a badly named .dll file.
Now he're's another
```
Exception: System.IO.FileNotFoundException: Could not load file or assembly 'file:///C:\Users\Administrator\Desktop\Halcyon\bin\OpenSim.Region.ClientStack.LindenUDP.dll' or one of its dependencies. The system cannot find the file specified.
File name: 'file:///C:\Users\Administrator\Desktop\Halcyon\bin\OpenSim.Region.ClientStack.LindenUDP.dll'
```
But there is a .dll file in there called OpenSim.Region.ClientStack.dll

Any relation?

**Comment by kf6kjg** [comment]
No, no relation.

A proper compile will have all of the following similarly named DLLs along with many others:
* Halcyon.Data.Inventory.MySQL.dll
* OpenSim.Data.dll
* OpenSim.Data.MySQL.dll
* OpenSim.Data.Null.dll
* OpenSim.Region.ClientStack.dll
* OpenSim.Region.ClientStack.LindenUDP.dll

No DLLs should need any renaming.  What was the name before you renamed it?

**Comment by emperorstarfinder** [comment]
Seeing as I have been watching this issue this evening, I quickly 
grabbed a copy of the latest Halcyon code and did a test build.

Unfortunately, I could not reproduce this particular issue.  However, it 
does sound like Visual Studios is not properly getting the files.  This 
could be due to a couple of reasons:

a) The user has attempted to clone a copy of the repository from within 
Visual Studios itself and it did not obtain a clean copy (this sometimes 
can happen but rare)
b) There is an issue with the solution files again, as this does seem to 
be similar to issues relating to renaming files (you shouldn't need to 
do this unless you are rebranding for your own purposes) I had found 
prior to Ricky updating the solution files
c) Github is having issues causing a clean copy of the repository being 
downloaded either through cloning in a local git repository software 
(i.e Tortoisegit, Smartgit, etc) (this would be rare but it has been 
known to happen before)
d) The user has an unclean installation of Visual Studio or does not 
have the necessary frameworks or Windows SDKs installed (I install all 
.Net Framework and Windows SDKs to ensure builds work properly)

The solutions, in this case, would be

a) Download a compressed folder of the master branch from the Halcyon 
Repository and attempt to build locally
b) CHeck to be sure all necessary.Net Framework and Windows SDKs are 
installed

Unfortunately, this does not appear to be a code problem that I can see.

Hope this helps a little bit.

**Comment by DJShocker** [comment]
emperorstarfinder,
Thank you. I took your advice and just downloaded the zip from GitHub and those errors are gone. Just now failed to load Whip Server, but I think I can take it from here. Thanks, I'm new to all of this and doing this just to teach myself a little something.

**Comment by DJShocker** [comment]
Actually, Whip doesn't have prebuilts and only instructions for Linux.

**Comment by kf6kjg** [comment]
Actually WHIP has compilation instructions for both Windows and Ubuntu Linux: https://github.com/HalcyonGrid/whip-server/wiki/Compiling-from-source

Glad Emporer was able to help you find a way to get compiled. Those were very strange issues.

**Comment by DJShocker** [comment]
It says build.cmd but I think it meant build.bat?

**Comment by kf6kjg** [comment]
Invited DJShocker onto the Discord and we continued resolving setup support there.
